src/include/btree_inline.h

```diff
@@ -2065,45 +2128,84 @@ __wt_btcur_skip_page(
-     * page. We base this decision on the aggregate stop point added to the page during the last
-     * reconciliation. We can skip this test if the page has been modified since it was reconciled.
-     * We also skip this test on an internal page, as we rely on reconciliation to mark the internal
-     * page dirty. There could be a period of time when the internal page is marked clean but the
-     * leaf page is dirty and has newer data than let on by the internal page's aggregated
-     * information.
+     * page.
      *
-     * We are making these decisions while holding a lock for the page as checkpoint or eviction can
-     * make changes to the data structures (i.e., aggregate timestamps) we are reading. It is okay
-     * if the page is not in memory, or gets evicted before we lock it. In such a case, we can forgo
-     * checking if the page has been modified. So, only do a page modified check if the page was in
-     * memory before locking.
+     * Skip this test on an internal page, as we rely on reconciliation to mark the internal page
+     * dirty. There could be a period of time when the internal page is marked clean but the leaf
+     * page is dirty and has newer data than let on by the internal page's aggregated information.
      */
     if (F_ISSET(ref, WT_REF_FLAG_INTERNAL))
         return (0);
 
+    /*
+     * We are making these decisions while holding a lock for the page as checkpoint or eviction can
+     * make changes to the data structures (i.e., aggregate timestamps) we are reading.
+     */
     WT_REF_LOCK(session, ref, &previous_state);
-    if ((previous_state == WT_REF_DISK || previous_state == WT_REF_DELETED ||
-          (previous_state == WT_REF_MEM && !__wt_page_is_modified(ref->page))) &&
-      __wt_ref_addr_copy(session, ref, &addr) && addr.ta.newest_stop_txn != WT_TXN_MAX &&
-      addr.ta.newest_stop_ts != WT_TS_MAX &&
-      __wt_txn_visible(session, addr.ta.newest_stop_txn, addr.ta.newest_stop_ts))
+
+    /*
+     * Check the fast-truncate information, there are 4 cases:
+     *
+     * (1) The page is in the WT_REF_DELETED state and ft_info.del is NULL. The page is deleted.
+     * (2) The page is in the WT_REF_DELETED state and ft_info.del is not NULL. The page is deleted
+     *     if the truncate operation is visible. Look at ft_info.del; we could use the info from the
+     *     address cell below too, but that's slower.
+     * (3) The page is in the WT_REF_DISK state. The page may be deleted; check the delete info from
+     *     the address cell.
+     * (4) The page is in memory and has been instantiated. The delete info from the address cell
+     *     will serve for readonly/unmodified pages, and for modified pages we can't skip the page
+     *     anyway.
+     */
+    if (previous_state == WT_REF_DELETED &&
+      (ref->ft_info.del == NULL ||
+        __wt_txn_visible(session, ref->ft_info.del->txnid, ref->ft_info.del->timestamp))) {
```

Comment:
Should we check prepared here as well?

Reply:
I don't think it's necessary? But I'm not sure. If the truncate's transaction is prepared, it isn't committed, which means it shouldn't be visible in the snapshot we're using. Unless we're doing read-uncommitted but in that case we probably expect to see the truncate. But I might be overlooking something.

I'm not sure why __wt_page_del_visible/active has a special case for prepare; I'm not convinced it's necessary, but I also didn't want to change it without evidence that it was wrong.